src/ert/storage/local_ensemble.py

```diff
@@ -885,73 +909,118 @@ def get_observations_and_responses(
-                    pivoted,
-                    by=["response_key", *response_cls.primary_key],
-                    on="time",
-                    tolerance="1s",
-                )
-            else:
-                joined = observations_for_type.join(
-                    pivoted,
-                    how="left",
-                    on=["response_key", *response_cls.primary_key],
+        with polars.StringCache():
+            dfs_per_response_type = []
+            for (
+                response_type,
+                response_cls,
+            ) in self.experiment.response_configuration.items():
+                if response_type not in observations_by_type:
+                    continue
+
+                observations_for_type = (
+                    observations_by_type[response_type]
+                    .filter(
+                        polars.col("observation_key").is_in(list(selected_observations))
+                    )
+                    .with_columns(
+                        [
+                            polars.col("response_key")
+                            .cast(polars.Categorical)
+                            .alias("response_key")
+                        ]
+                    )
                 )
 
-            joined = (
-                joined.with_columns(
-                    polars.concat_str(response_cls.primary_key, separator=", ").alias(
-                        "__tmp_index_key__"  # Avoid potential collisions w/ primary key
+                observed_cols = {
+                    k: observations_for_type[k].unique()
+                    for k in ["response_key", *response_cls.primary_key]
+                }
+
+                reals = iens_active_index.tolist()
+                reals.sort()
+                # too much memory to do it all at once, go per realization
+                first_columns: polars.DataFrame | None = None
+                realization_columns: List[polars.DataFrame] = []
+                for real in reals:
+                    my_responses = self._load_responses_lazy(
+                        response_type, (real,)
+                    ).with_columns(
+                        [
+                            polars.col("response_key")
+                            .cast(polars.Categorical)
+                            .alias("response_key")
+                        ]
                     )
-                )
-                .drop(response_cls.primary_key)
-                .rename({"__tmp_index_key__": "index"})
-            )
 
-            first_columns = [
-                "response_key",
-                "index",
-                "observation_key",
-                "observations",
-                "std",
-            ]
-            joined = joined.select(
-                first_columns + [c for c in joined.columns if c not in first_columns]
-            )
+                    # Filter out irrelevant data before reading into memory
```

Comment:
Add some context to what is irrelevant? I.e. response where there is no observation

Reply:
Fixed